src/plugins/defoldcollection/defoldcollectionplugin.cpp

```diff
@@ -0,0 +1,473 @@
+  \"  id: \\\"{{tilemap_name}}\\\"\\n\"\n\
+  \"  component: \\\"{{tilemap_rel_path}}\\\"\\n\"\n\
+  \"  position {\\n\"\n\
+  \"    x: 0.0\\n\"\n\
+  \"    y: 0.0\\n\"\n\
+  \"    z: 0.0\\n\"\n\
+  \"  }\\n\"\n\
+  \"  rotation {\\n\"\n\
+  \"    x: 0.0\\n\"\n\
+  \"    y: 0.0\\n\"\n\
+  \"    z: 0.0\\n\"\n\
+  \"    w: 1.0\\n\"\n\
+  \"  }\\n\"\n\
+  \"}\\n\"\n\
+";
+
+static const char child_t[] =
+"  children: \"{{child-name}}\"\n\
+";
+
+static const char emdedded_instance_t[] =
+"embedded_instances {\n\
+  id: \"{{instance-name}}\"\n\
+  data: {{components}}\n\
+  \"\"\n\
+  position {\n\
+    x: 0.0\n\
+    y: 0.0\n\
+    z: 0.0\n\
+  }\n\
+  rotation {\n\
+    x: 0.0\n\
+    y: 0.0\n\
+    z: 0.0\n\
+    w: 1.0\n\
+  }\n\
+  scale3 {\n\
+    x: 1.0\n\
+    y: 1.0\n\
+    z: 1.0\n\
+  }\n\
+}\n\
+";
+
+static QString replaceTags(QString context, const QVariantHash &map)
+{
+    QHashIterator<QString,QVariant> it{map};
+    while (it.hasNext()) {
+        it.next();
+        context.replace(QLatin1String("{{") + it.key() + QLatin1String("}}"),
+                        it.value().toString());
+    }
+    return context;
+}
+
+DefoldCollectionPlugin::DefoldCollectionPlugin()
+{
+}
+
+QString DefoldCollectionPlugin::nameFilter() const
+{
+    return tr("Defold collection (*.collection)");
+}
+
+QString DefoldCollectionPlugin::shortName() const
+{
+    return QLatin1String("defoldcollection");
+}
+
+QString DefoldCollectionPlugin::errorString() const
+{
+    return mError;
+}
+
+/*
+ * Returns a new filepath relative to the root of the Defold project if we're in one.
+ * Determines the root of the project by looking for a file called "game.project".
+ * If no such file is found by going up the heirarchy, return filename from the filepath.
+*/
+QString DefoldCollectionPlugin::TilesetRelativePath(QString filePath)
+{
+    QString gameproject = "/game.project";
+    QFileInfo fi(filePath);
+    QDir Qd = fi.dir();
+
+    while (Qd.exists() && !Qd.isRoot())
+    {
+        if (QFileInfo::exists(Qd.path() + gameproject))
+        {
+            // return relative path
+            filePath.replace(Qd.path(), "");
+            return filePath;
+        }
+        else if (!Qd.cdUp())
+        {
+            break;
+        }
+    }
+    return fi.fileName();
+}
+
+bool DefoldCollectionPlugin::write(const Tiled::Map *map, const QString &fileName)
+{
+    QFileInfo fi(fileName);
+    QString outputFilePath = fi.filePath();
+    QString outputFileName = fi.fileName();
+
+    QString mapName = outputFileName;
+    mapName.replace(".collection", "");
+
+    QString tilesetFileDir = outputFilePath;
+    tilesetFileDir.chop(outputFileName.length());
+
+    // if there are no group layers, all tilemaps are components of the top gameobject called "tilemaps"
+    if (map->groupLayerCount() == 0)
+    {
+        QString components;
+
+        // create a tilemap for each tileset this map uses
+        for (auto& tileset_current : map->tilesets())
```

Comment:
Coding style: please don't use underscores in variable names. Can be renamed to simply `tileset` I think. Also, please place the `&` against the variable name instead of the type.